Detectors/MUON/MCH/Raw/Decoder/src/DataDecoder.cxx

```diff
@@ -418,23 +462,30 @@ void DataDecoder::decodePage(gsl::span<const std::byte> page)
   uint32_t feeId;
   uint32_t linkId;
 
-  /*
-   * TODO: we should use the HBPackets to verify the synchronization between the SAMPA chips
   auto heartBeatHandler = [&](DsElecId dsElecId, uint8_t chip, uint32_t bunchCrossing) {
-    SampaInfo sampaId;
-    sampaId.chip = chip;
-    sampaId.ds = dsElecId.elinkId();
-    sampaId.solar = dsElecId.solarId();
+    auto ds = dsElecId.elinkId();
+    auto solar = dsElecId.solarId();
+    uint64_t chipId = getChipId(solar, ds, chip);
+
+    uint32_t bcTF = subtractBcOffset(bunchCrossing, mSampaTimeOffset);
+
+    updateTimeFrameStartRecord(chipId, mFirstOrbitInTF, bcTF);
 
     if (mDebug) {
       auto s = asString(dsElecId);
-      LOGP(info, "HeartBeat: {} ID {} SOLAR {} ds {} chip {} -> bxcount {} orbit {} [ bc {} ]",
-           s, sampaId.id, csampaId.solar, csampaId.ds, csampaId.chip,
-           bunchCrossingCounter, mOrbit,
-           bunchCrossingCounter % 3564);
+      LOGP(info, "HeartBeat: {}-CHIP{} -> {}/{} offset {} -> bcTF {}",
+           s, chip, mFirstOrbitInTF, bunchCrossing, mSampaTimeOffset, bcTF);
+    }
+
+    if (!mTimeFrameStartRecords[chipId].check()) {
+      if (mErrorCount < MCH_DECODER_MAX_ERROR_COUNT) {
+        auto s = asString(dsElecId);
+        LOGP(warning, "Inconsistent HeartBeat packet received: {}-CHIP{} {}/{} (last {}/{})",
```

Comment:
I guess this is fine for now, but cutting out messages might be dangerous in the long run. A better option (even if it'll cost some CPU) would be to store messages, counting their occurrences, and showing that store once in a while. (might require to think a bit in order not to generate too different messages for similar errors, otherwise the message store would blow up the memory...)